examples/llm/src/data/denoising.py

```diff
@@ -0,0 +1,705 @@
+                prefix_tokens = prefix_function(sequence_mask_ratio, None,
+                                                self.tokenizer)
+            else:
+                prefix_tokens = None
+
+            max_raw_length = self.max_seq_length - len(prefix_tokens or []) - 1
+
+            if not self._uses_span_corruption and (
+                    max_raw_length < self._smallest_max_raw_length):
+                # We choose not to count sequence denoising in the smallest
+                # unless there is only sequence denoising.
+                self._smallest_max_raw_length = max_raw_length
+            if max_raw_length > self._largest_max_raw_length:
+                self._largest_max_raw_length = max_raw_length
+
+            kwargs = {
+                'mean_span_length': None,
+                'mask_ratio': sequence_mask_ratio,
+                'prefix_tokens': prefix_tokens,
+                'max_raw_length': max_raw_length,
+            }
+            self._noisers.append(kwargs)
+
+        if not self._noisers:
+            raise ValueError(
+                'No denoising tasks were included. Make sure to set ' + \
+                '`span_mean_lengths_and_ratios` and/or `sequence_mask_ratios`.')
+
+    @property
+    def smallest_max_raw_length(self):
+        return int(self._smallest_max_raw_length)
+
+    @property
+    def largest_max_raw_length(self):
+        return int(self._largest_max_raw_length)
+
+    def __call__(self, examples: List[Dict[str,
+                                           Any]]) -> Dict[str, torch.Tensor]:
+        """Batch examples processed by the span corrupter."""
+        processed_examples = []
+        for example in examples:
+            # Randomly pick a "noiser" to apply to this example
+            noiser = random.choice(self._noisers)
+            # Apply it
+            processed_examples.append(
+                noise_token_sequence(
+                    example,
+                    mask_ratio=noiser['mask_ratio'],
+                    mean_span_length=noiser['mean_span_length'],
+                    prefix_tokens=noiser['prefix_tokens'],
+                    max_raw_length=noiser['max_raw_length'],
+                    max_seq_length=self.max_seq_length,
+                    tokenizer=self.tokenizer,
+                    sentinel_token_ids=self._sentinel_token_ids,
+                    decoder_only_format=self.decoder_only_format))
+        batch = self.tokenizer.pad(processed_examples)
+
+        # Truncate portions of the inputs that are purely padding
+        # (up to a multiple of 8)
+        multiple_of = 8
+        n_examples_per_length = batch['attention_mask'].sum(0)
+        keep_tokens = torch.sum(n_examples_per_length > 0)
+        keep_tokens = int(multiple_of * torch.ceil(keep_tokens / multiple_of))
+
+        # Note: EncDec formatting will always produce a right-padded batch
+        if self.tokenizer.padding_side == 'left' and self.decoder_only_format:
```

Comment:
is there any case where we want to support left-padding? I think it presents a bit of a challenge for our causal LM with certain attention implementations, but then again I guess this denoiser will never be used by causal LMs...

Reply:
My intuition was that the collator should not override the padding side choice (except in the weird case where someone tries to use a left-padded-by-default tokenizer, for example, with EncDec formatting).

You're correct that this collator shouldn't be used with causal LMs, but if there's a causal LM that can't support a given padding side, its prefix LM version will have the same limitation. The good news is that I think HF has been careful to match the default padding side of a tokenizer with the expected padding side of any model implementation. I think that, by supporting both padding sides, this collator helps to take advantage of that.